polygeist/tools/cgeist/Lib/ValueCategory.cc

```diff
@@ -44,9 +44,15 @@ ValueCategory::ValueCategory(mlir::Value val, bool isReference)
 
 ValueCategory::ValueCategory(mlir::Value Val, mlir::Value Index)
     : val{Val}, isReference{true}, Index{Index} {
-  assert(val.getType().isa<MemRefType>() &&
-         val.getType().cast<MemRefType>().getElementType().isa<VectorType>() &&
-         "Expecting memref of vector");
+  assert(
+      ((val.getType().isa<MemRefType>() &&
+        val.getType().cast<MemRefType>().getElementType().isa<VectorType>()) ||
+       (val.getType().isa<LLVM::LLVMPointerType>() &&
+        val.getType()
+            .cast<LLVM::LLVMPointerType>()
+            .getElementType()
+            .isa<VectorType>())) &&
```

Comment:
`!llvm.ptr<vector>` is not created from `getMLIRType`. 
`sycl::detail::Boolean` is represented as `struct<vector>`,  where pointer to struct is represented using `llvm.ptr`.
When accessing the vector, it uses `llvm.getelementptr`, which produces a `llvm.ptr`, e.g., `%0 = llvm.getelementptr %arg1[0, 0] : (!llvm.ptr<struct<(vector<4xi8>)>, 4>) -> !llvm.ptr<vector<4xi8>, 4>`.

Reply:
If we implement `sycl::detail::Boolean` in the SYCL dialect, pointer to it is represented using `memref`.
When accessing the vector, it would use `polygeist.subindex`, which produces a `memref`.
That's why we have memref of vector in `sycl.vec` examples, and pointer of vector in this case (`sycl::detail::Boolean`).